src/devices/Qmc5883l/Qmc5883l.cs

```diff
@@ -0,0 +1,150 @@
+using System.Buffers.Binary;
+using System.Device.I2c;
+using System.Device.Model;
+using System.IO;
+using System.Numerics;
+using UnitsNet;
+
+namespace Iot.Device.Qmc5883l
+{
+    /// <summary>
+    /// 3-Axis Digital Compass QMC5883L
+    /// </summary>
+    [Interface("3-Axis Digital Compass QMC5883L")]
+    public class Qmc5883l : IDisposable
+    {
+        /// <summary>
+        /// QMC5883L Default I2C Address
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x0D;
+        private I2cDevice _i2cDevice;
+        private bool _isDeviceInitialized = false;
+
+        /// <summary>
+        /// Enables or disables the Interrupt PIN.
+        /// </summary>
+        /// <value>Disabled by default.</value>
+        public Interrupt Interrupt { get; set; } = Interrupt.Disable;
+
+        /// <summary>
+        /// When the point roll-over function is enabled, the I2C data pointer automatically rolls between 00H ~ 06H.
+        /// </summary>
+        /// <value>Disabled by default.</value>
+        public RollPointer RollPointer { get; set; } = RollPointer.Disable;
+
+        /// <summary>
+        /// Typical Data Output Rate (Hz).
+        /// </summary>
+        /// <value>10Hz by default.</value>
+        public OutputRate OutputRate { get; set; } = OutputRate.Rate10Hz;
+
+        /// <summary>
+        /// Bandwidth of an internal digital filter.
+        /// </summary>
+        /// <value>256 by default.</value>
+        public Oversampling Oversampling { get; set; } = Oversampling.Rate256;
+
+        /// <summary>
+        /// The field range goes hand in hand with the sensitivity of the magnetic sensor.
+        /// </summary>
+        /// <value>8 Gauss by default.</value>
+        public FieldRange FieldRange { get; set; } = FieldRange.Gauss8;
+
+        /// <summary>
+        /// Sensors operation mode.
+        /// </summary>
+        /// <value>In Standby mode by default. Make sure to change it to Continuous before trying to get any data!</value>
+        public Mode DeviceMode { get; set; } = Mode.Standby;
+
+        /// <summary>
+        /// QMC5883L Direction Vector
+        /// </summary>
+        [Telemetry]
+        public Vector3 GetDirection() => _isDeviceInitialized ? ReadDirectionVector() : throw new IOException("Sensor has not yet been initialized. Please call SetMode() to initialize it.");
+
+        /// <summary>
+        /// QMC5883L Heading (DEG)
+        /// </summary>
+        /// <returns>Heading(Angle)</returns>
+        public Angle GetHeading() => VectorExtension.GetHeading(GetDirection());
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Qmc5883l"/> class.
+        /// </summary>
+        public Qmc5883l(I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentNullException(nameof(i2cDevice));
+        }
+
+        /// <summary>
+        /// Sets the sensors mode.
+        /// </summary>
+        public void SetMode()
```

Comment:
this should be an implementation detail

Reply:
i.e. call from the setter of each property. I recommend to update/read a flags related to a mode you're changing without any extra variables directly in the getter and setter